RIOT-OS/RIOT
# gnrc_icmpv6_error: fix and use where appropriate

### Contribution description
When implementing ICMPv6 for GNRC I also added build and send functions for ICMPv6 error messages, but I never actually used them while implementing IPv6. Reviewing that sub-module I found it quite buggy:

* Packets in receive order were assumed in send order
* Packets were sent out without an IP header attached
* Packets were just sent to the PID of the IP thread instead of searching the registry
* etc.

This PR fixes those issues. Additionally a call of the send functions is added were appropriate, so GNRC is now able to correctly (but optionally) report to a sender what went wrong. To prevent further degeneration of that module, I activated it for `gnrc_networking` (but also some tests might be nice, which is why I set the WIP label, though the tests can also come as a follow-up IMHO).

### Testing procedures

I tested this using `examples/gnrc_networking` and [scapy](https://scapy.readthedocs.io/en/latest/). If you get problems receiving you might need to merge #10356 first. The tests are required to be executed in order as commands previously executed in other tests might be a precondition:

1. Compile and start `gnrc_networking` (make sure `gnrc_icmpv6_error` is included; should be automatically as per this PR) for native
    ```sh
    make -C examples/gnrc_networking all term
    ``` 
2. Get the link-local IPv6 address and hardware address of the native instance:
    ```
    > ifconfig
   Iface  6  HWaddr: 36:54:72:4F:8B:1F 
              MTU:1500  HL:64  RTR  
              RTR_ADV  
              Source address length: 6
              Link type: wired
              inet6 addr: fe80::3454:72ff:fe4f:8b1f  scope: local  VAL
              inet6 group: ff02::2
              inet6 group: ff02::1
              inet6 group: ff02::1:ff4f:8b1f
              inet6 group: ff02::1a
              
              Statistics for Layer 2
                RX packets 32  bytes 6964
                TX packets 28 (Multicast: 6)  bytes 6980
                TX succeeded 28 errors 0
              Statistics for IPv6
                RX packets 32  bytes 6516
                TX packets 28 (Multicast: 6)  bytes 6588
                TX succeeded 28 errors 0
    ```
3. Get the link-local IPv6 address of your TAP interface/bridge
4. Start `scapy` with root privileges (required to send ethernet frames)
5. Set some variables:
   ```py
   DST_ADDR = "<native link-local IPv6 address>"
   DST_HWADDR = "<native hardware address>"
   SRC_ADDR = "<TAP interface link-local IPv6 address>"
   ```
6. Test UDP "port unreachable" error:
    1. Port not registered:
          **scapy:**
          ```py
          p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst=DST_ADDR, src=SRC_ADDR) / UDP(),
                    iface="tapbr0")
          assert(ICMPv6DestUnreach in p)
          assert(p[ICMPv6DestUnreach].code == 4) # Port unreachable
          ```
    2. Port registered:
          **RIOT:**
          ```py
         udp server start 1337
          ```
          **scapy:**
          ```py
          p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst=DST_ADDR, src=SRC_ADDR) / UDP(dport=1337),
                    iface="tapbr0", timeout=1)
          assert(p == None)
          ```
7. Check if large packets are carried correctly:
    ```py
    # 1452 = 1500 (Ethernet payload length) - 40 (IPv6 header size) - 8 (UDP header size)
    p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst=DST_ADDR, src=SRC_ADDR) / UDP() / (1452 * "x"),
                   iface="tapbr0")
    assert(ICMPv6DestUnreach in p)
    assert(p[ICMPv6DestUnreach].code == 4) # Port unreachable
    assert(len(p[Raw]) < 1452)
    ```
8. Test "address unreachable" (local address):
    ```py
    p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst="fe80::1", src=SRC_ADDR) / UDP(), iface="tapbr0", timeout=1)
    assert(ICMPv6DestUnreach in p)
    assert(p[ICMPv6DestUnreach].code == 3) # Address unreachable
    ```
9. Test "beyond scope of source address":
    ```py
   p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst="affe::1", src=SRC_ADDR) / UDP(), iface="tapbr0", timeout=1)
    assert(ICMPv6DestUnreach in p)
    assert(p[ICMPv6DestUnreach].code == 2) # Beyond scope of source address
    ```
10. Test "erroneous header field encountered":
    ```py
    p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst=DST_ADDR, src=SRC_ADDR, plen=20) / UDP(),
                   iface="tapbr0")
    assert(ICMPv6ParamProblem in p)
    assert(p[ICMPv6ParamProblem].code == 0) # erroneous header field encountered
    assert(p[ICMPv6ParamProblem].ptr == 4)  # IPv6 length field
    ```
11. Test "No route to destination"
    **Linux**: Configure route to native instance and set a global source address:
    ```sh
    sudo ip a a beef::1 dev tapbr0
    sudo ip r a affe::/64 via "<native link-local IPv6 address>" dev tapbr0
    ```
    **RIOT**: Configure route back to Linux:
    ```
    nib route add 6 beef::/64 "<TAP interface link-local IPv6 address>"
    ```
    **scapy**:
    ```py
    p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst="affe::1", src="beef::1") / UDP(), iface="tapbr0")
    assert(ICMPv6DestUnreach in p)
    assert(p[ICMPv6DestUnreach].code == 0) # No route to destination
    ```
12. Test "Address unreachable" (neighboring node):
    **RIOT**: Configure route to bogus neighbor:
    ```
    nib route add 6 affe::/64 fe80::1
    ```
    **scapy**:
    ```py
    sendp(Ether(dst=DST_HWADDR) / IPv6(dst="affe::1", src="beef::1") / UDP(), iface="tapbr0")
    # srp1 would not return, for some reason
    # just check with wireshark or a second `scapy` instance using
    ps = sniff(iface="tapbr0", count=20)
    p = [p for p in ps if ICMPv6DestUnreach in p]
    assert(len(p) == 1)
    assert(p[0][ICMPv6DestUnreach].code == 3) # Address unreachable
    ```
13. Test "Hop limit exceeded in transit"
    ```py
    p = srp1(Ether(dst=DST_HWADDR) / IPv6(dst="affe::1", src="beef::1", hlim=1) / UDP(), iface="tapbr0")
    assert(ICMPv6TimeExceeded in p)
    assert(p[ICMPv6TimeExceeded].code == 0) # Hop limit exceeded in transit
    ```

### Issues/PRs references
None
